### PR TITLE
blog/2024: Ceph: A Journey to 1 TiB/s

### PR DESCRIPTION
**Pull request by markhpc**
First Draft is done!

### CODE REVIEW

File: src/en/news/blog/2024/ceph-a-journey-to-1tibps/index.md [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
...irst 3 node test looked good.  With 10 OSDs per node, We were achieving roughly proportional (IE higher) performance relative to the previous tests.  I knew I wasn't going to have much time to do proper scaling tests, so I immediately bumped up from 3 nodes to 10 nodes. I also scaled the PG count at the same time and used CBT to deploy a new cluster.  At 3 nodes I saw 63GiB/s for 4MB random reads.  At 10 Nodes, I saw 213.5GiB/s.  That's almost linear scaling at 98.4%.  It was at this point that I knew that things were finally taking a turn for the better. Of the 68 nodes for this cluster, only 63 were up at that time.  The rest were down for maintenance to fix various issues.  I split the cluster roughly in half, with 32 nodes (320 OSDs) in one half, and 31 client nodes running 10 FIO process each in the other half.  I watched as CBT built the cluster over roughly a 7-8 minute period of time.  The initial write prefill looked really good.  My heart soared.  We were reading data at 635 GiB/s.  We broke 15 million 4k random read IOPS.  While this may not seem impressive compared to the individual NVMe drives, these were the highest numbers I had ever seen for a ~300 OSD Ceph cluster.
```
Comment (anthonyeleven) [inline src/en/news/blog/2024/ceph-a-journey-to-1tibps/index.md:228]: s/recreate/repeat/
io_depth in backticks
I.e.
Suggest a space between 63 GiB/s here and elsewhere

File: src/en/news/blog/2024/ceph-a-journey-to-1tibps/index.md [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+
+![](images/Post-Fixes_OSD_Scaling_-_FIO_4MB_Throughput.svg)
+![](images/Post-Fixes_OSD_Scaling_-_FIO_4KB_IOPS.svg)
+
+I also plotted both average and tail latency for the scaling tests.  Both looked consistent. This was likely due to scaling the PG count and the FIO client count at the same time as OSDs.  These tests are very IO heavy however.  We have so much client traffic that we are likely well into the inflection point where performance doesn't increase while latency continues to grow as more IO is added.
```
Comment (anthonyeleven) [inline src/en/news/blog/2024/ceph-a-journey-to-1tibps/index.md:233]: s/consistent/constant/ ?

File: src/en/news/blog/2024/ceph-a-journey-to-1tibps/index.md [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+
+![](images/Post-Fixes_OSD_Scaling_-_FIO_4KB_Average_Latency.svg)
+![](images/Post-Fixes_OSD_Scaling_-_FIO_4KB_99%_Latency.svg)
+
+I showed these results to my colleague Dan van der Ster who previously had built the Ceph infrastructure at CERN.  He bet me a beer (Better be a good one Dan!) if I could hit 1 TiB/s.  I told him that had been my plan since the beginning.
+
+# A Partially Operational Death Star
+
+I had no additional client nodes to test the cluster with at full capacity, so the only real option was to co-locate FIO processes on the same nodes as the OSDs.  On one hand, this provides a very slight network advantage.  Clients will be able to communicate with local OSDs 1/63rd of the time.  On the other hand, we know from previous testing that co-locating fio clients on OSD nodes isn't free.  There's often a performance hit, and it wasn't remotely clear to me how much of a hit a cluster of this scale would take.
```
Comment (anthonyeleven) [inline src/en/news/blog/2024/ceph-a-journey-to-1tibps/index.md:242]: FIO

File: src/en/news/blog/2024/ceph-a-journey-to-1tibps/index.md [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+![](images/Post-Fixes_OSD_Scaling_-_FIO_4KB_Average_Latency.svg)
+![](images/Post-Fixes_OSD_Scaling_-_FIO_4KB_99%_Latency.svg)
+
+I showed these results to my colleague Dan van der Ster who previously had built the Ceph infrastructure at CERN.  He bet me a beer (Better be a good one Dan!) if I could hit 1 TiB/s.  I told him that had been my plan since the beginning.
+
+# A Partially Operational Death Star
+
+I had no additional client nodes to test the cluster with at full capacity, so the only real option was to co-locate FIO processes on the same nodes as the OSDs.  On one hand, this provides a very slight network advantage.  Clients will be able to communicate with local OSDs 1/63rd of the time.  On the other hand, we know from previous testing that co-locating fio clients on OSD nodes isn't free.  There's often a performance hit, and it wasn't remotely clear to me how much of a hit a cluster of this scale would take. 
+
+I built a new CBT configuration targeting the 63 nodes I had available.  Deploying the cluster with CBT took about 15 minutes to stand up all 630 OSDs and build the pool.  I waited with baited breath and watched the results as they came in.
```
Comment (anthonyeleven) [inline src/en/news/blog/2024/ceph-a-journey-to-1tibps/index.md:244]: bated

File: src/en/news/blog/2024/ceph-a-journey-to-1tibps/index.md [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+
+![](images/Full_Cluster_Initial_Tuning_-_FIO_4MB_Throughput.svg)
+![](images/Full_Cluster_Initial_Tuning_-_FIO_4KB_IOPS.svg)
+
+Around 950GiB/s.  So very very close.  It was late on Friday night at this point, so I wrapped up and turned in for the night.  On Saturday morning I logged in and threw a couple of tuning options at the cluster:  Lowering OSD shards and async messenger threads while also applying the Reef RocksDB tunings. As you can see, we actually hurt read performance a little while helping write performance.  In fact random write performance improved by nearly 20%.  After further testing it looked like the reef tunings were benign though only helping a little bit in the write tests.  The bigger effect seemed to be coming from shard/thread changes.  At this point I had to take a break and wasn't able to get back to working on the cluster again until Sunday night.  I tried to go to bed, but I knew that I was down to the last 24 hours before we needed to wrap this up.  At around midnight I gave up on sleep and got back to work.
```
Comment (anthonyeleven) [inline src/en/news/blog/2024/ceph-a-journey-to-1tibps/index.md:249]: Reef

File: src/en/news/blog/2024/ceph-a-journey-to-1tibps/index.md [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+
+Despite being distracted by the laggy PG issue, I wanted to refocus on hitting 1.0TiB/s.  Sleep was finally catching up with me.  At some point I had doubled the PG count again to 256K, just to see if it had any affect at all on the laggy PG issue.  That puts us solidly toward the upper end of the curve we showed earlier, though frankly I don't think it actually mattered much.  I decided to switch back to the default OSD shard counts and continue testing with 504 fio client processes.  I did however scale the number of async messenger threads.  There were two big takeaways.  The first is that dropping down to 1 async messenger allowed us to avoid PGs going laggy and achieve “OK” write throughput with 504 clients.  It also dramatically hurt 4MB read performance.  The second:  Ceph's defaults were actually ideal for 4MB reads.  With 8 shards, 2 threads per shard, and 3 msgr threads, we finally broke 1TiB/s.  Here's the view I had at around 4AM Monday morning as the final set of tests for the night ran:
```
Comment (anthonyeleven) [inline src/en/news/blog/2024/ceph-a-journey-to-1tibps/index.md:278]: Lack of sleep

That put us solidly
FIO

File: src/en/news/blog/2024/ceph-a-journey-to-1tibps/index.md [hunk 7]
```diff
@@ -0,0 +1,345 @@
...
+```
+
+and the graphs from the fio results:
+
+![](images/Full_Cluster_Msgr_Thread_Scaling_-_FIO_4MB_Throughput.svg)
+![](images/Full_Cluster_Msgr_Thread_Scaling_-_FIO_4KB_IOPS.svg)
+
+# Sleep; Erasure Coding
+
+After finally seeing the magical "1.0 TiB/s" screen I had been wait weeks to see, I finally went to sleep.  Nevertheless, I got up several hours later.  There was still work to be done.  All of the testing we had done so far was on 3X replication, but the customer would be migrating this hardware into an existing cluster deployed with 6+2 erasure coding.  We needed to get some idea of what this cluster was capable of in the configuration they would be using.
```
Comment (anthonyeleven) [inline src/en/news/blog/2024/ceph-a-journey-to-1tibps/index.md:304]: s/was on/was with/

File: src/en/news/blog/2024/ceph-a-journey-to-1tibps/index.md [hunk 8]
```diff
@@ -0,0 +1,345 @@
...
+It's not clear yet if we will be able to get this merged for Squid, though Radoslaw Zarzynski, core lead for the Ceph project, has offered to help try to get this over the finish line.
+
+# Squeezing in Msgr Encryption Testing
+
+Finally, we wanted to provide the customer with a rough idea how much msgr level encryption would impact their cluster if they decided to use it.  The adrenaline of the previous night had long faded and I was dead tired at this point.  I managed to run through both 3X replication and 6+2 erasure coding tests with msgr v2 encryption enabled and compared it against our previous test results.
+
+![](images/Full_Cluster_Msgr_Encryption_-_FIO_4MB_Throughput.svg)
+![](images/Full_Cluster_Msgr_Encryption_-_FIO_4KB_IOPS.svg)
+
+The biggest hit is to large reads.  They drop from ~1 TiB/s to around 750 GiB/s.  Everything else sees a more modest, though consistent hit.  At this point I had to stop.  I really wanted to do PG scaling tests and even kernel RBD tests, but it was time to hand the systems back to the customer for re-imaging, followed by one of my excellent colleagues at Clyso to begin the integration process.
```
Comment (anthonyeleven) [inline src/en/news/blog/2024/ceph-a-journey-to-1tibps/index.md:327]: at Clyso beginning the integration

File: src/en/news/blog/2024/ceph-a-journey-to-1tibps/index.md [hunk 9]
```diff
@@ -0,0 +1,345 @@
...
+
+# Finale
+
+So what's happened with this cluster since the end of the testing? All hardware was re-imaged and the new OSDs were deployed into the customer's existing HDD cluster.  Dan's upmap-remapped script is being used to control the migration process and we've migrated around 80% of the existing data to the NVMe backed OSDs.  By next week, the cluster should be fully migrated to the new NVMe based nodes.  We've opted not to employ all of the tuning we've done here, at least not at first.  Initially we'll make sure the cluster behaves well under the existing, mostly default configuration.  We now have a mountain of data we can use to tune the system further if the customer hits any performance issues.
+
+Since there was a ton of data and a ton of graphs here, I want to recap some of the highlights.  Here's an outline of the best numbers we were able to achieve on this cluster:
+
+|                | 30 OSDs (3x) | 100 OSDs (3x) | 320 OSDs (3x) | 630 OSDs (3x) | 630 OSDs (EC62)
+| -------------- | ------------ | ------------- | ------------- | ------------- | ---------------
+| Co-Located Fio | No           | No            | No            | Yes           | Yes
```
Comment (anthonyeleven) [inline src/en/news/blog/2024/ceph-a-journey-to-1tibps/index.md:337]: FIO

File: src/en/news/blog/2024/ceph-a-journey-to-1tibps/index.md [hunk 10]
```diff
@@ -0,0 +1,345 @@
...
+| 4KB Rand Write | 248K IOPS    | 745K IOPS     | 2.4M IOPS     | 4.9M IOPS     | 936K IOPS
+
+What's next? We need to figure out how to fix the laggy PG issue during writes.  We can't have Ceph falling apart when the write workload gets too big.  Beyond that, we learned through this exercise that Ceph is perfectly capable of saturating 2x100GbE NICs.  To push the throughput envelope further we will need 200GbE+ when using 10 NVMe drives per node or more.  IOPS is more nuanced. We know that PG count can have a big effect.  We also know that the general OSD threading model is playing a big role.  We consistently hit a wall at around 400-600K random read IOPS per node and we've seen it in multiple deployments.  Part of this may be how the async msgr interfaces with the kernel and part of this may be how OSD threads wake up when new work is put into the shard queues.  I've modified the OSD code in the past to achieve better results under heavy load, but at the expense of low-load latency.  Ultimately, I suspect improving IOPS will take a multi-pronged approach and a rewrite of some of the OSD threading code.
```
Comment (anthonyeleven) [inline src/en/news/blog/2024/ceph-a-journey-to-1tibps/index.md:343]: s/gets too big/scales up/
space after 2x -- suggest adding up at the top whether there was a separate physical or VLAN replication network, and if bonding was used, and with which hash policy

**Review by anthonyeleven — APPROVED** [review]

File: src/en/news/blog/2024/ceph-a-journey-to-1tibps/index.md [hunk 11]
```diff
@@ -0,0 +1,345 @@
...
+
+# Acknowledgments
+
+Thank you first to our amazing customer that made all of this possible.  You were a pleasure to work with!  Thank you as well for allowing us here at [Clyso](https://www.clyso.com) to share this experience with the Ceph community.  It is through this sharing of knowledge that we make the world a better place.  Thank you as well to [IBM](https://ibm.com)/[Red Hat](https://redhat.com) and [Samsung](https://samsung.com) for providing the Ceph community with the hardware used for comparison testing.  It was invaluable to be able to evaluate the numbers we were getting against previous tests from the lab.  Thank you as well to all of the Ceph developers who have worked tirelessly to make Ceph great!
```
[the remark this answers is not in the document]
Comment (markhpc) [reply src/en/news/blog/2024/ceph-a-journey-to-1tibps/index.md:24]: Absolutely!

File: src/en/news/blog/2024/ceph-a-journey-to-1tibps/index.md [hunk 12]
```diff
@@ -0,0 +1,345 @@
...
+
+# Cluster Setup 
+
+When the customer first approached Clyso, they proposed a configuration utilizing 34 dual-socket 2U nodes spread across 17 racks.  We provided a couple of alternative configurations from multiple vendors with a focus on smaller nodes.  Ultimately they decided to go with a Dell architecture we designed that quoted out roughly 13% cheaper than the original configuration despite having several key advantages.  The new configuration has less memory per OSD (still comfortably 12GiB each), but faster memory throughput.  It also provides more aggregate CPU resources, significantly more aggregate network throughput, a simpler single-socket configuration, and utilizes the newest generation of AMD processors and DDR5 RAM.  By employing smaller nodes, we halved the impact of a node failure on cluster recovery.
```
[the remark this answers is not in the document]
Comment (markhpc) [reply src/en/news/blog/2024/ceph-a-journey-to-1tibps/index.md:28]: It's tough to keep up to date.  This was the first Genoa system we've been able to test.  I don't have much info on modern Xeon with DDR5 either so I'll be curious to hear how it goes if you go that route!

File: src/en/news/blog/2024/ceph-a-journey-to-1tibps/index.md [hunk 13]
```diff
@@ -0,0 +1,345 @@
...
+
+The customer indicated they would like to limit the added per-rack power consumption to around 1000-1500 watts.  With 4 of these nodes per rack, the aggregate TDP is estimated to be at least 1120 Watts plus base power usage, CPU overage peaks, and power supply inefficiency.  IE it's likely we're pushing it a bit under load, but we don't expect significant deviation beyond the acceptable range.  If worse came to worst, we estimated we could shave off roughly 100 watts per rack by lowering the processor cTDP.
```
[the remark this answers is not in the document]
Comment (markhpc) [reply src/en/news/blog/2024/ceph-a-journey-to-1tibps/index.md:30]: Yeah, we had to do that at the supercomputing institute I worked at after turning one of our new clusters on blew the breakers!

File: src/en/news/blog/2024/ceph-a-journey-to-1tibps/index.md [hunk 14]
```diff
@@ -0,0 +1,345 @@
...
+CPU | 1 x AMD EPYC 9454P 48C/96T
+Memory | 192GiB DDR5
+Network | 2 x 100GbE Mellanox ConnectX-6
+NVMe | 10 x Dell 15.36TB Enterprise NVMe Read Intensive AG
+OS Version | Ubuntu 20.04.6 (Focal)
+Ceph Version | Quincy v17.2.7 (Upstream Deb Packages)
+
+An additional benefit of utilizing 1U Dell servers is that they are essentially a newer refresh of the systems David Galloway and I designed for the upstream Ceph performance lab.  These systems have been tested in a [variety](https://ceph.io/en/news/blog/2023/ceph-encryption-performance/) of [articles](https://ceph.io/en/news/blog/2023/reef-osds-per-nvme/) over the past couple of years.  It turns out that there was a major performance impacting issue that came out during testing that did not affect the previous generation of hardware in the upstream lab, but did affect this new hardware.  We'll talk about that more later.
+
+Without getting into too many details, I will simply reiterate that the customer's network configuration is very well designed and quite fast.  It easily has enough aggregate throughput across all 17 racks to let a cluster of this scale really stretch it's legs.
```
[the remark this answers is not in the document]
Comment (markhpc) [reply src/en/news/blog/2024/ceph-a-journey-to-1tibps/index.md:45]: done!

File: src/en/news/blog/2024/ceph-a-journey-to-1tibps/index.md (hunk 6 above)
Comment (markhpc) [reply src/en/news/blog/2024/ceph-a-journey-to-1tibps/index.md:278]: Done!  Also fixed other fio->FIO spots.

File: src/en/news/blog/2024/ceph-a-journey-to-1tibps/index.md (hunk 7 above)
Comment (markhpc) [reply src/en/news/blog/2024/ceph-a-journey-to-1tibps/index.md:304]: Done!

File: src/en/news/blog/2024/ceph-a-journey-to-1tibps/index.md (hunk 8 above)
Comment (markhpc) [reply src/en/news/blog/2024/ceph-a-journey-to-1tibps/index.md:327]: Didn't like how that sounded.  How about this?

"...but it was time to hand the systems back to the customer for re-imaging. Once completed, one of my excellent colleagues at Clyso would begin the integration process."

File: src/en/news/blog/2024/ceph-a-journey-to-1tibps/index.md [hunk 15]
```diff
@@ -0,0 +1,345 @@
...
+
+# Testing Setup
+
+To do the burn-in testing, ephemeral ceph clusters were deployed and FIO tests were launched using [CBT](https://github.com/ceph/cbt/).  CBT was configured to deploy Ceph with several modified settings.  OSDs were assigned an 8GB osd_memory_target, however in production a higher osd_memory_target should be acceptable.  The customer had no need to test block or S3 workloads, so one might assume that RADOS bench would be the natural benchmark choice.  In my experience, testing at large scale with RADOS bench is tricky.  It's tough to determine how many instances are needed to saturate the cluster at given thread counts.  I've run into issues in the past where multiple concurrent pools were needed to scale performance.  I also didn't have any preexisting RADOS bench tests handy to compare against.  Instead, we opted to do burn-in testing using the same librbd backed FIO testing we've used in the upstream lab.  This allowed us to partition the cluster into smaller chunks and compare results with previously published results.  FIO is also very well known both inside and outside the Ceph community and is well trusted.
```
[the remark this answers is not in the document]
Comment (markhpc) [reply src/en/news/blog/2024/ceph-a-journey-to-1tibps/index.md:49]: done!

File: src/en/news/blog/2024/ceph-a-journey-to-1tibps/index.md (hunk 9 above)
Comment (markhpc) [reply src/en/news/blog/2024/ceph-a-journey-to-1tibps/index.md:337]: done!

File: src/en/news/blog/2024/ceph-a-journey-to-1tibps/index.md (hunk 10 above)
Comment (markhpc) [reply src/en/news/blog/2024/ceph-a-journey-to-1tibps/index.md:343]: Done for the first part.  For the second part, I'm going to leave that out for now.  If the customer wants to expose more details regarding their network configuration we can go back and add it as an addendum.  I just want to be a little cautious about sharing details of things we don't have explicit permission to share.

File: src/en/news/blog/2024/ceph-a-journey-to-1tibps/index.md (hunk 8 above)
Comment (anthonyeleven) [reply src/en/news/blog/2024/ceph-a-journey-to-1tibps/index.md]: What about "for re-imaging, then to one of my excellent colleagues at Clyso for integration" ?

Comment (markhpc) [reply src/en/news/blog/2024/ceph-a-journey-to-1tibps/index.md]: I can role with that

**Pull request merged**